### ISSUE TITLE
Measure the performance of each analyzer in the ruling test

### ISSUE THREAD

**Issue by dbolkensteyn** [SonarSource/sonarlint-vs#722]
This issue depends on #664 to be implemented first.

The C# compiler, `csc.exe`, has an option `/reportanalyzer`, controlled by `$(ReportAnalyzer)` in MSBuild, to report the execution time of every analyzer in this format:
```
Total analyzer execution time: 0.385 seconds.
NOTE: Elapsed time may be less than analyzer execution time because analyzers can run concurrently.

Time (s)    %   Analyzer
   0.385  100   SonarLint.CSharp, Version=1.6.0.0, Culture=neutral, PublicKeyToken=null
   0.019    4      SonarLint.Rules.CSharp.RedundantModifier
   0.018    4      SonarLint.Rules.CSharp.ExpressionComplexity
   0.016    4      SonarLint.Rules.CSharp.StaticFieldInGenericClass
   0.012    3      SonarLint.Rules.CSharp.SingleStatementPerLine
   0.012    3      SonarLint.Rules.CSharp.DisposableNotDisposed
   0.010    2      SonarLint.Rules.CSharp.DisposeNotImplementingDispose
   0.009    2      SonarLint.Rules.CSharp.TooManyLabelsInSwitch
   0.008    2      SonarLint.Rules.CSharp.DisposableMemberInNonDisposableClass
   0.008    2      SonarLint.Rules.CSharp.UseCurlyBraces
   0.008    2      SonarLint.Rules.CSharp.ParametersCorrectOrder
   0.008    1      SonarLint.Rules.CSharp.CommentedOutCode
   0.007    1      SonarLint.Rules.CSharp.ArgumentSpecifiedForCallerInfoParameter
   0.007    1      SonarLint.Rules.CSharp.MemberShadowsOuterStaticMember
   0.007    1      SonarLint.Rules.CSharp.ToStringNoNull
   0.006    1      SonarLint.Rules.CSharp.CollectionQuerySimplification
   0.006    1      SonarLint.Rules.CSharp.MethodOverloadOptionalParameter
   0.006    1      SonarLint.Rules.CSharp.HardcodedIpAddress
   0.006    1      SonarLint.Rules.CSharp.AsyncAwaitIdentifier
   0.006    1      SonarLint.Rules.CSharp.FunctionComplexity
   0.006    1      SonarLint.Rules.CSharp.FieldShouldBeReadonly
   0.006    1      SonarLint.Rules.CSharp.VariableShadowsField
   0.006    1      SonarLint.Rules.CSharp.VariableUnused
   0.006    1      SonarLint.Rules.CSharp.ClassWithOnlyStaticMember
   0.006    1      SonarLint.Rules.CSharp.GenericTypeParameterUnused
   0.005    1      SonarLint.Rules.CSharp.BooleanCheckInverted
   0.005    1      SonarLint.Rules.CSharp.CommentTodo
   0.004    1      SonarLint.Rules.CSharp.MethodParameterUnused
   0.004    1      SonarLint.Rules.CSharp.StringOperationWithoutCulture
   0.004    1      SonarLint.Rules.CSharp.PropertyGetterWithThrow
   0.004   <1      SonarLint.Rules.CSharp.RedundantArgument
   0.004   <1      SonarLint.Rules.CSharp.InvocationResolvesToOverrideWithParams
   0.004   <1      SonarLint.Rules.CSharp.GetHashCodeMutable
   0.004   <1      SonarLint.Rules.CSharp.StringFormatArgumentNumberMismatch
   0.003   <1      SonarLint.Rules.CSharp.EmptyNestedBlock
   0.003   <1      SonarLint.Rules.CSharp.CatchRethrow
   0.003   <1      SonarLint.Rules.CSharp.SuppressFinalizeUseless
   0.003   <1      SonarLint.Rules.CSharp.ForLoopCounterChanged
   0.003   <1      SonarLint.Rules.CSharp.EmptyMethod
   0.003   <1      SonarLint.Rules.CSharp.GetTypeWithIsAssignableFrom
   0.003   <1      SonarLint.Rules.CSharp.ConditionalSimplification
   0.003   <1      SonarLint.Rules.CSharp.RedundantCast
   0.002   <1      SonarLint.Rules.CSharp.OrderByRepeated
   0.002   <1      SonarLint.Rules.CSharp.AsyncVoidMethod
   0.002   <1      SonarLint.Rules.CSharp.MethodOverrideNoParams
   0.002   <1      SonarLint.Rules.CSharp.DisposeFromDispose
   0.002   <1      SonarLint.Rules.CSharp.ConstructorOverridableCall
   0.002   <1      SonarLint.Rules.CSharp.RedundantInheritanceList
   0.002   <1      SonarLint.Rules.CSharp.UseValueParameter
   0.002   <1      SonarLint.Rules.CSharp.StreamReadStatement
   0.002   <1      SonarLint.Rules.CSharp.PublicMethodWithMultidimensionalArray
   0.002   <1      SonarLint.Rules.CSharp.ConsoleLogging
   0.002   <1      SonarLint.Rules.CSharp.OptionalParameter
   0.002   <1      SonarLint.Rules.CSharp.ParameterNamesInPartialMethod
   0.002   <1      SonarLint.Rules.CSharp.EnumerableSumInUnchecked
   0.002   <1      SonarLint.Rules.CSharp.ArrayCovariance
   0.002   <1      SonarLint.Rules.CSharp.StaticFieldWrittenFromInstanceMember
   0.002   <1      SonarLint.Rules.CSharp.GuardConditionOnEqualsOverride
   0.002   <1      SonarLint.Rules.CSharp.CollectionEmptinessChecking
   0.002   <1      SonarLint.Rules.CSharp.ReferenceEqualsOnValueType
   0.002   <1      SonarLint.Rules.CSharp.BooleanLiteralUnnecessary
   0.002   <1      SonarLint.Rules.CSharp.StringConcatenationInLoop
   0.002   <1      SonarLint.Rules.CSharp.MultilineBlocksWithoutBrace
   0.002   <1      SonarLint.Rules.CSharp.FieldInitializedToDefault
   0.002   <1      SonarLint.Rules.CSharp.SillyBitwiseOperation
   0.002   <1      SonarLint.Rules.CSharp.GetHashCodeEqualsOverride
   0.002   <1      SonarLint.Rules.CSharp.GenericReadonlyFieldPropertyAssignment
   0.001   <1      SonarLint.Rules.CSharp.TooManyParameters
   0.001   <1      SonarLint.Rules.CSharp.RedundantParentheses
   0.001   <1      SonarLint.Rules.CSharp.ParameterAssignedTo
   0.001   <1      SonarLint.Rules.CSharp.DelegateSubtraction
   0.001   <1      SonarLint.Rules.CSharp.IfConditionalAlwaysTrueOrFalse
   0.001   <1      SonarLint.Rules.CSharp.GotoStatement
   0.001   <1      SonarLint.Rules.CSharp.ReversedOperators
   0.001   <1      SonarLint.Rules.CSharp.CastConcreteTypeToInterface
   0.001   <1      SonarLint.Rules.CSharp.GenericTypeParameterInOut
   0.001   <1      SonarLint.Rules.CSharp.PropertyWriteOnly
   0.001   <1      SonarLint.Rules.CSharp.ConditionalStructureSameCondition
   0.001   <1      SonarLint.Rules.CSharp.IndexOfCheckAgainstZero
   0.001   <1      SonarLint.Rules.CSharp.EqualityOnFloatingPoint
   0.001   <1      SonarLint.Rules.CSharp.AssignmentInsideSubExpression
   0.001   <1      SonarLint.Rules.CSharp.VirtualEventField
   0.001   <1      SonarLint.Rules.CSharp.ObjectCreatedDropped
   0.001   <1      SonarLint.Rules.CSharp.ConditionalStructureSameImplementation
   0.001   <1      SonarLint.Rules.CSharp.SelfAssignment
   0.001   <1      SonarLint.Rules.CSharp.ForLoopCounterCondition
   0.001   <1      SonarLint.Rules.CSharp.ThreadStaticWithInitializer
   0.001   <1      SonarLint.Rules.CSharp.PropertyToAutoProperty
   0.001   <1      SonarLint.Rules.CSharp.LockOnThisOrType
   0.001   <1      SonarLint.Rules.CSharp.AnonymousDelegateEventUnsubscribe
   0.001   <1      SonarLint.Rules.CSharp.UnaryPrefixOperatorRepeated
   0.001   <1      SonarLint.Rules.CSharp.ShortCircuitNullPointerDereference
   0.001   <1      SonarLint.Rules.CSharp.ForeachLoopExplicitConversion
   0.001   <1      SonarLint.Rules.CSharp.GenericTypeParameterEmptinessChecking
   0.001   <1      SonarLint.Rules.CSharp.CatchEmpty
   0.001   <1      SonarLint.Rules.CSharp.DisposableReturnedFromUsing
   0.001   <1      SonarLint.Rules.CSharp.MultipleVariableDeclaration
   0.001   <1      SonarLint.Rules.CSharp.DefaultBaseConstructorCall
   0.001   <1      SonarLint.Rules.CSharp.DeadStores
   0.001   <1      SonarLint.Rules.CSharp.LiteralSuffixUpperCase
   0.001   <1      SonarLint.Rules.CSharp.BinaryOperationWithIdenticalExpressions
   0.001   <1      SonarLint.Rules.CSharp.StaticFieldInitializerOrder
   0.001   <1      SonarLint.Rules.CSharp.FlagsEnumWithoutInitializer
   0.001   <1      SonarLint.Rules.CSharp.ValuesUselesslyIncremented
   0.001   <1      SonarLint.Rules.CSharp.TernaryOperatorPointless
   0.001   <1      SonarLint.Rules.CSharp.FlagsEnumZeroMember
   0.001   <1      SonarLint.Rules.CSharp.IfCollapsible
   0.001   <1      SonarLint.Rules.CSharp.EnumNameHasEnumSuffix
   0.001   <1      SonarLint.Rules.CSharp.TaskConfigureAwait
   0.001   <1      SonarLint.Rules.CSharp.SwitchWithoutDefault
   0.001   <1      SonarLint.Rules.CSharp.StaticFieldVisible
   0.001   <1      SonarLint.Rules.CSharp.FieldShadowsParentField
  <0.001   <1      SonarLint.Rules.CSharp.ThreadStaticNonStaticField
  <0.001   <1      SonarLint.Rules.CSharp.InsecureEncryptionAlgorithm
  <0.001   <1      SonarLint.Rules.CSharp.EmptyStatement
  <0.001   <1      SonarLint.Rules.CSharp.CommentFixme
  <0.001   <1      SonarLint.Rules.CSharp.ClassShouldNotBeAbstract
  <0.001   <1      SonarLint.Rules.CSharp.InsecureHashAlgorithm
  <0.001   <1      SonarLint.Rules.CSharp.EqualityOnModulus
  <0.001   <1      SonarLint.Rules.CSharp.RightCurlyBraceStartsLine
  <0.001   <1      SonarLint.Rules.CSharp.TabCharacter
  <0.001   <1      SonarLint.Rules.CSharp.MethodName
  <0.001   <1      SonarLint.Rules.CSharp.FileLines
  <0.001   <1      SonarLint.Rules.CSharp.IfChainWithoutElse
  <0.001   <1      SonarLint.Rules.CSharp.FieldShouldNotBePublic
  <0.001   <1      SonarLint.Rules.CSharp.ClassName
  <0.001   <1      SonarLint.Rules.CSharp.SwitchCasesMinimumThree
  <0.001   <1      SonarLint.Rules.CSharp.LineLength
  <0.001   <1      SonarLint.Rules.CSharp.BreakOutsideSwitch
  <0.001   <1      SonarLint.Rules.CSharp.PublicConstantField
  <0.001   <1      SonarLint.Rules.CSharp.ConditionalsWithSameCondition
```

To see this, the compiler must be called with the `/reportanalyzer` flag.

This particular output was obtained using the following command:
```
csc /t:exe /out:Foo.exe Program.cs /analyzer:C:\Users\Dinesh\Desktop\tmp\ConsoleApplication1\.sonarqube\bin\SonarLint.CSharp.dll /reportanalyzer
```

on the following `Program.cs` file:
```
// TODO FIXME
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ConsoleApplication1
{
    static class Program
    {
        static void Main()
        {
            ;
            ;

            Console.WriteLine("waiting for input...");
            Console.ReadLine();

            return;
        }
    }
}
```

One would need to look at the default C# targets file to see how the `/reportanalyzer` flag can be passed to CSC through MSBuild.

**Comment by dbolkensteyn** [comment]
Note: I don't think we need to hardcode any assertion for the time being... it's just nice to have the chance to manually look at these numbers...
